architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,329 @@
+| Normal resource type | A resource type that is part of the `radius` plane and exists inside a resource group. Also has the `TrackedResource` set of fields and behaviors. This is the case for most resource types in Radius today. Normal resource types are our scenario priority because they are optimized for being part of an application. The resource types that are not *normal* are not usually part of an application, for example `System.AWS/credentials` is not part of an application and does not exist in a resource group. |
+
+## Objectives
+
+
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6688
+
+### Guiding principles
+
+We have two guiding principles for how we make decisions and prioritize investments. 
+- User-defined types must be simple and declarative for application and recipe scenarios.  The highest value scenario is building custom application platforms. We want to build a big ecosystem and so user-defined types should be easy to adopt and easy to develop.
+- User-defined types should be powerful enough for the Radius maintainers to use. We want to leverage user-defined types and declarative functionality as much as possible.
+- 
+### Goals
+
+- Support users defining their own resource types and resource type schemas for use with Radius.
+	- Support rich declarative validation of resource request bodies based on schema.
+	- Support versioning for user-defined resource types.
+		- Including the ability to support multiple API versions.
+		- Including the ability to introduce breaking changes between versions.
+	- eg: Users can define a new resource type and it's schema. Radius will use the schema for validation. Users can add new versions of the resource type, and old ones continue to function.
+- Support dynamic management of resource types.
+	- eg: Users can register and unregister resource types.
+	- eg: Tooling like the Dashboard and `rad` CLI dynamically query the set of resource types rather than hardcoding a list of known types.
+- Support use of user-defined types in Bicep with full fidelity.
+	- eg: Users can define their own types and use them in their Bicep application definitions.
+- Support use of user-defined types as part of an application with full fidelity.
+	- User defined types can opt-in to support connections, recipes, application-graph.
+	- eg: Users can define their own types and incorporate them in applications. They work the same as existing types that weren't defined by the user.
+- Support multiple implementation choices for user-defined types.
+	- User-defined types should be easy to implement using declarative functionality.
+	- User-defined types should be powerful enough to support implementing a resource-provider.
```

Comment:
Could you share scenario(s) you had on mind that would motivate users to implement a new resource provider?

Reply:
I think the key scenario is the existing functionality in Radius - it's already been written. Another is *special* types like `Applications.Core/secretStores` (for it's handling of secrets).

I hope the community won't have to write full RPs, its too much work.